### PR TITLE
Normalizemodels

### LINKED ISSUE #31: Create REST GET endpoint for all users
Using [Django REST framework](https://www.django-rest-framework.org/), create a GET endpoint to query all users. 

GET /users/

### LINKED ISSUE #32: Create a GET endpoint for a single user
Using [Django REST framework](https://www.django-rest-framework.org/), create a GET endpoint to query a single user by id. 

GET /users/<id>

### LINKED ISSUE #33: Create a POST endpoint for a single user
Using [Django REST framework](https://www.django-rest-framework.org/), create a POST endpoint to create a single user. 

POST /users/<id>

### LINKED ISSUE #34: Add PUT/PATCH functionality for user/<id> endpoint
Using [Django REST framework](https://www.django-rest-framework.org/), create a PUT/PATCH endpoint to edit a single user by id. 

PUT /users/<id>
PATCH /users/<id>

### LINKED ISSUE #35: DELETE single user by id
Using [Django REST framework](https://www.django-rest-framework.org/), create a DELETE endpoint to delete a single user by id. 

DELETE /users/<id>

### PR DESCRIPTION
**Pull request by Nkoli**
Fixes #31 
Fixes #32 
Fixes #33 
Fixes #34 
Fixes #35

### CODE REVIEW

File: .gitignore
Comment (Del-sama) [inline .gitignore]: Please add a line at the end

File: app/admin.py
Comment (Del-sama) [inline app/admin.py]: Why was this commented out?

File: app/migrations/0010_auto_20200420_1833.py
Comment (Del-sama) [inline app/migrations/0010_auto_20200420_1833.py:1]: What's the difference between this migration and the migration before it?

File: app/migrations/0011_auto_20200420_2003.py
Comment (Del-sama) [inline app/migrations/0011_auto_20200420_2003.py:3]: See above comment

File: app/serializers.py
Comment (Del-sama) [inline app/serializers.py:55]: This isn't necessary, but you can include all fields by passing `__all__` to fields this `fields = '__all__'`. https://www.django-rest-framework.org/api-guide/serializers/#modelserializer

File: app/serializers.py
Comment (Del-sama) [inline app/serializers.py]: Why are you overriding this method?

File: app/views/language.py
Comment (Del-sama) [inline app/views/language.py:7]: Why did you opt for Generic views? 
Is this like a default view file? I see that all the classes do the same things

File: app/views/user.py
Comment (Del-sama) [inline app/views/user.py]: I don't know if you want to use a viewSet here. They don't implement the regular API CRUD operations like get and post

**Review by Del-sama — COMMENTED** [review]
This looks good, Nkoli. Good job!
Please go through my comments.

**Comment by Del-sama** [comment]
When this is approved, it'll be nice to rebase the commits before you merge.

**Comment by Nkoli** [comment]
> This looks good, Nkoli. Good job!
> Please go through my comments.

File: app/views/language.py
Comment (Nkoli) [reply app/views/language.py:7]: Generic views is modeled like a template for requests, with APIView, I'd have had to write out the GET request function and I didn't think that was necessary since I don't need to customize the language proficiency. ListAPIView only handles get all objects request.

File: app/serializers.py
Comment (Nkoli) [reply app/serializers.py]: Found that this method allowed me create the skill and language data as you suggested last two weeks I think (name: str, proficiency_id: itspredefinedid).

File: app/serializers.py
Comment (Nkoli) [reply app/serializers.py:55]: Yea I know, I'm not too sure why I don't like passing all the fields that way though lol.

File: .gitignore
Comment (Nkoli) [reply .gitignore]: Okay

File: app/admin.py
Comment (Nkoli) [reply app/admin.py]: Allowed me display user_id in the django-admin page (before I got started with the API). Commented it out when I found I didn't need it anymore since it shows up because of the serializer.

File: app/migrations/0011_auto_20200420_2003.py
Comment (Nkoli) [reply app/migrations/0011_auto_20200420_2003.py:3]: You advised I add the unique attribute so that one doesn't need to add the same values to the table over and over.

File: app/views/user.py
Comment (Nkoli) [reply app/views/user.py]: I was able to use it to create and get users though. I'm not sure I understand, please explain? It's specifically because I was trying to use the user_id (not drf's default id) for retrieve/put/patch/delete operations that I even used viewset, also because I'm unable to properly write the logic to use user_id as the lookup id.

File: app/serializers.py
Comment (Del-sama) [reply app/serializers.py:55]: That's fine also.

File: app/serializers.py
Comment (Del-sama) [reply app/serializers.py]: I'll have to look into how the method works, I think I could create my views without overriding it.

File: app/views/user.py
Comment (Del-sama) [reply app/views/user.py]: I'll have to look into this also.

File: app/serializers.py
Comment (Nkoli) [reply app/serializers.py]: Okay. I did use your project as a guide but i noticed most of the foreign keys were somewhat dependent on the parent object existing (then the foreign key can be passed in as an id).

File: app/views/user.py
Comment (Nkoli) [reply app/views/user.py]: Okay.

File: app/serializers.py
Comment (kosyfrances) [inline app/serializers.py]: You'd need to re-order imports in the following order:
```
Standard library imports.
Related third party imports.
Local application/library specific imports.
```
 https://www.python.org/dev/peps/pep-0008/#imports

File: app/models.py
Comment (kosyfrances) [inline app/models.py:172]: I think we can make this help_text better. No worries about it though, just a note. We can always fix it and all the other help_texts later.

File: app/serializers.py
Comment (kosyfrances) [inline app/serializers.py]: Would you like to keep these commented serialisers? Or delete them?

File: app/admin.py
Comment (kosyfrances) [reply app/admin.py]: Would you like to remove the commented lines then?

File: app/urls.py
Comment (kosyfrances) [inline app/urls.py:16]: We don't need a new line before this import

File: app/models.py
Comment (kosyfrances) [inline app/models.py:40]: Does the cloudinary field work with the current API setup? Or does this need to be commented out also, since it was commented out in the settings?

File: app/admin.py
Comment (Nkoli) [reply app/admin.py]: mm, not yet though. Maybe when we're all done with the backend?

File: app/serializers.py
Comment (Nkoli) [reply app/serializers.py]: Okay, will do.

File: app/models.py
Comment (Nkoli) [reply app/models.py:172]: Okay, till then. Even the README.md file needs some edits.

File: app/urls.py
Comment (Nkoli) [reply app/urls.py:16]: I know, but it looks cleaner this way.

File: app/models.py
Comment (Nkoli) [reply app/models.py:40]: well it has not thrown any errors so far, however through postman, I can't directly upload an image so I just append .png or .jpg file to a name when I'm testing. I'll uncomment it out in the settings, didn't know it was commented out

File: app/serializers.py
Comment (Nkoli) [reply app/serializers.py]: Tempted to remove them, at least Anthony can have a fresh start, that sort of thing.

File: app/admin.py
Comment (kosyfrances) [reply app/admin.py]: You can always add it back, or make a note to not forget about it though. Your call :)

File: app/models.py
Comment (kosyfrances) [reply app/models.py:40]: yea of course what it needs is the name from the API. pls can you check whether you can upload an image that way? If you can't, please just make a new ticket for it so we don't hold down this PR longer

File: app/models.py
Comment (kosyfrances) [reply app/models.py:172]: Pls make a new ticket for it.

File: app/serializers.py
Comment (kosyfrances) [reply app/serializers.py]: you can remove it and he can follow the pattern of what is already existing and understand what is going on

File: app/urls.py
Comment (kosyfrances) [reply app/urls.py:16]: okay

File: app/models.py
Comment (Nkoli) [reply app/models.py:172]: will do

File: app/serializers.py
Comment (Nkoli) [reply app/serializers.py]: okay then.

File: app/models.py
Comment (Nkoli) [reply app/models.py:40]: right then, will do

**Review by Del-sama — COMMENTED** [review]
I think you need to rebase your branch against staging

**Comment by Del-sama** [comment]
I rebased your branch against staging, please review to make sure the rebase is correct. If it is correct, delete your local branch and fetch this. If you need help with this, please let me know.

File: app/models.py
Comment (Del-sama) [reply app/models.py:40]: https://stackoverflow.com/questions/36805137/how-to-pass-options-to-cloudinaryfield-in-django-model

File: app/serializers.py
Comment (Del-sama) [inline app/serializers.py]: duplicate RoleSerializer

File: app/serializers.py
Comment (Del-sama) [inline app/serializers.py]: Same here

File: app/serializers.py
Comment (Del-sama) [inline app/serializers.py]: The `None` here isn't necessary, I think. Python's `dict.get()` method returns `None` when the key doesn't exist. So this can be `validated_data.get('role')`

File: app/views/user.py
Comment (Del-sama) [inline app/views/user.py:32]: I think you should return a response here instead of raising an error.  `return Response(f"user with user id : {pk} does not exist", status=status.HTTP_404_NOT_FOUND)`

File: app/views/user.py
Comment (Del-sama) [inline app/views/user.py:50]: Just to explain why this is a 204 https://restfulapi.net/http-methods/

File: app/views/user.py
Comment (Del-sama) [inline app/views/user.py:39]: `PATCH method is the correct choice for partially updating an existing resource and PUT should only be used if you’re replacing a resource in its entirety`. Turns out PATCH might be the correct thing here

Comment (Del-sama) [reply app/views/user.py:39]: I think we should leave this as is though.

```HTTP PATCH

HTTP PATCH requests are to make partial update on a resource. If you see PUT requests also modify a resource entity so to make more clear – PATCH method is the correct choice for partially updating an existing resource and PUT should only be used if you’re replacing a resource in its entirety.

Please note that there are some challenges if you decide to use PATCH APIs in your application:

    Support for PATCH in browsers, servers, and web application frameworks is not universal. IE8, PHP, Tomcat, Django, and lots of other software has missing or broken support for it.
    Request payload of PATCH request is not straightforward as it is for PUT request. e.g.

    HTTP GET /users/1

    produces below response:

    {id: 1, username: 'admin', email: 'email@example.org'}

    A sample patch request to update the email will be like this:

    HTTP PATCH /users/1
    [
    { “op”: “replace”, “path”: “/email”, “value”: “new.email@example.org” }
    ]

There may be following possible operations are per the HTTP specification.

[
{ "op": "test", "path": "/a/b/c", "value": "foo" },
{ "op": "remove", "path": "/a/b/c" },
{ "op": "add", "path": "/a/b/c", "value": [ "foo", "bar" ] },
{ "op": "replace", "path": "/a/b/c", "value": 42 },
{ "op": "move", "from": "/a/b/c", "path": "/a/b/d" },
{ "op": "copy", "from": "/a/b/d", "path": "/a/b/e" }
]

PATCH method is not a replacement for the POST or PUT methods. It applies a delta (diff) rather than replacing the entire resource.

```

**Review by Del-sama — CHANGES_REQUESTED** [review]
Looks good asides the cloudinary thing and the comments I left. I think whatever we allow into staging shouldn't have commented out code.

File: app/views/user.py
Comment (Nkoli) [reply app/views/user.py:50]: thank you

**Comment by Del-sama** [comment]
LGTM. Please squash your commits before you merge. I can help with that and show you how to go about it later.

**Review by Del-sama — APPROVED** [review]

**Comment by kosyfrances** [comment]
@Nkoli if it is not too much of a hassle, can you please squash the migration files into one? (if)

**Pull request merged**
